### PR TITLE
Iox #350 implement listener c binding

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md

## Notes for Reviewer
-  The Listener C-Binding is implemented and documented (examples will follow with the next PR)
- Adjustments in the `cxx::vector`, added `resize` - required by the listener unit tests.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

### CODE REVIEW

File: iceoryx_binding_c/include/iceoryx_binding_c/listener.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+#include "iceoryx_binding_c/internal/c2cpp_binding.h"
+#include "iceoryx_binding_c/subscriber.h"
+#include "iceoryx_binding_c/types.h"
+#include "iceoryx_binding_c/user_trigger.h"
+
+typedef CLASS Listener* iox_listener_t;
+
+
+/// @brief initializes a listener struct from a storage struct pointer
+/// @param self pointer to raw memory which can hold a listener
```
Comment (dkroenke) [inline iceoryx_binding_c/include/iceoryx_binding_c/listener.h:30]: /// param [in/out] for all doxygen comments here

File: iceoryx_binding_c/include/iceoryx_binding_c/types.h [hunk 2]
```diff
@@ -66,4 +67,13 @@ struct iox_pub_storage_t_
 };
 typedef struct iox_pub_storage_t_ iox_pub_storage_t;
 
+struct iox_listener_storage_t_
+{
+    // the value of the array size is the result of the following formula:
+    // sizeof(WaitSet) / 8
+    uint64_t do_not_touch_me[2567];
```
Comment (dkroenke) [inline iceoryx_binding_c/include/iceoryx_binding_c/types.h:74]: is here some safety margin in or strictly tied to the size?

File: iceoryx_binding_c/source/cpp2c_subscriber.cpp [hunk 3]
```diff
@@ -33,7 +33,16 @@ void cpp2c_Subscriber::enableEvent(iox::popo::TriggerHandle&& triggerHandle,
     static_cast<void>(subscriberEvent);
 
     m_trigger = std::move(triggerHandle);
-    iox::popo::SubscriberPortUser(m_portData).setConditionVariable(m_trigger.getConditionVariableData());
+    if (triggerHandle.doesContainEventVariable())
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_binding_c/source/cpp2c_subscriber.cpp:36]: Please update the copyright header.

File: iceoryx_utils/include/iceoryx_utils/cxx/vector.hpp [hunk 4]
```diff
@@ -38,54 +38,54 @@ class vector
     using const_iterator = const T*;
 
     /// @brief creates an empty vector
-    vector() = default;
+    vector() noexcept = default;
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_utils/include/iceoryx_utils/cxx/vector.hpp:41]: Please update the copyright header.

File: iceoryx_utils/test/moduletests/test_cxx_vector.cpp [hunk 5]
```diff
@@ -1116,3 +1116,99 @@ TEST_F(vector_test, PopBackReturnsTrueTillItsEmpty)
...
+        int m_a;
+    };
+    iox::cxx::vector<DefaultCTor, 10> sut;
+
+    EXPECT_TRUE(sut.resize(5));
+    for (auto& e : sut)
+    {
+        EXPECT_THAT(e.m_a, Eq(1231));
+    }
+    EXPECT_THAT(sut.size(), Eq(5));
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_utils/test/moduletests/test_cxx_vector.cpp:1148]: Maybe this should be checked before the for loop (and replaced by ASSERT_*). Consider this also for the following tests.

File: iceoryx_utils/test/moduletests/test_cxx_vector.cpp [hunk 6]
```diff
@@ -1116,3 +1116,99 @@ TEST_F(vector_test, PopBackReturnsTrueTillItsEmpty)
...
+}
+
+TEST_F(vector_test, SizeIncreaseWithResizeAndTemplateValueWorks)
+{
+    EXPECT_TRUE(sut.resize(4, 421337));
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_utils/test/moduletests/test_cxx_vector.cpp:1153]: I see a lot of magic numbers and I miss a lot of ``U``s ;)

**Review by FerdinandSpitzschnueffler — COMMENTED** [review]
I reviewed all files except ``test_listener.cpp`` - will continue tomorrow.

File: iceoryx_binding_c/test/moduletests/test_listener.cpp [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+    MemoryManager m_memoryManager;
+
+    const iox::capro::ServiceDescription TEST_SERVICE_DESCRIPTION{"a", "b", "c"};
+
+    iox::popo::SubscriberOptions subscriberOptions{MAX_CHUNKS_HELD_PER_SUBSCRIBER_SIMULTANEOUSLY, 0U};
+
+    cxx::vector<iox::popo::SubscriberPortData, MAX_NUMBER_OF_EVENTS_PER_LISTENER + 1> m_subscriberPortData;
+    cxx::vector<cpp2c_Subscriber, MAX_NUMBER_OF_EVENTS_PER_LISTENER + 1> m_subscriber;
+    cxx::vector<ChunkQueuePusher<SubscriberPortData::ChunkQueueData_t>, MAX_NUMBER_OF_EVENTS_PER_LISTENER + 1>
+        m_chunkPusher;
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_binding_c/test/moduletests/test_listener.cpp:156]: still missing

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Comment by dkroenke** [comment]
The issue is not linked.

**Comment by dkroenke** [comment]
unittests for `iox_listener_init` and `iox_listener_deinit` are missing, can this be tracked?

**Review by budrus — APPROVED** [review]

**Review by dkroenke — APPROVED** [review]

**Pull request merged**